File: ruby/core/lib/mullet/render_context.rb
require 'cgi'
require 'mullet/default_nested_scope'

module Mullet

  # Holds the rendering context to reduce the number of parameters passed to
  # render methods.
  class RenderContext

    attr_accessor :escape_xml_enabled

    # Constructor.
    #
    # @param [Object] data
    #           provides data to render
    # @param [Proc] missing_value_strategy
    #           executed on attempt to render a variable that was not found
    # @param [Proc] nil_value_strategy
    #           executed on attempt to render nil value
    # @param [#<<] output
    #           where to write rendered output
    def initialize(data, missing_value_strategy, nil_value_strategy, output)
      @scope = data.is_a?(DefaultNestedScope) ?
          data : DefaultNestedScope.new(data)
      @on_missing = missing_value_strategy
      @on_nil = nil_value_strategy
      @output = output
      @escape_xml_enabled = true
    end

    # Escapes characters that could be interpreted as XML markup if enabled.
    #
    # @param [String] input
    #           input string
    # @return escaped string, or the input string if escaping is disabled.
    def escape_xml(input)
      return @escape_xml_enabled ? CGI.escape_html(input.to_s()) : input
    end

    # Resolves variable name to value.
    #
    # @param [Symbol] name
    #           variable name
    # @return value
    def get_variable_value(name)
      return @scope.get_variable_value(name)
    end

    # Adds a nested scope to search in subsequent lookups.
    #
    # @param data
    #           data object
    def push_scope(data)
      @scope.push_scope(data)
    end

    # Removes innermost nested scope.
    def pop_scope()
      @scope.pop_scope()
    end

    # Gets scope value that is intended for display in the rendered output.
    # Applies configured strategies for handling missing and nil values.
    #
    # @param [Symbol] key
    #           variable name
    # @return value
    def get_display_value(key)
      value = @scope.get_variable_value(key)
      if value == Scope::NOT_FOUND
        value = @on_missing.call(key)
      end
      if value == nil
        value = @on_nil.call(key)
      end
      return value
    end

    # Writes rendered output.
    #
    # @param [String] str
    #           string to write
    # @return this object
    def <<(str)
      @output << str
      return self
    end
  end

end

File: java/core/src/main/java/com/github/pukkaone/mullet/RenderContext.java
/*
Copyright (c) 2011, Chin Huang
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.github.pukkaone.mullet;

import java.io.IOException;
import java.io.Writer;
import java.util.ResourceBundle;

/**
 * Holds the rendering context to reduce the number of parameters passed to
 * render methods.
 */
public class RenderContext {

    private NestedModel model;
    private ResourceBundle messages;
    private FailedValueStrategy missingValueStrategy;
    private FailedValueStrategy nullValueStrategy;
    private boolean escapeXmlEnabled = true;
    private Writer writer;

    public RenderContext(
            Object data,
            ResourceBundle messages,
            FailedValueStrategy missingValueStrategy,
            FailedValueStrategy nullValueStrategy, 
            Writer writer)
    {
        this.model = (data instanceof NestedModel)
                ? (NestedModel) data
                : new DefaultNestedModel(data);
        this.messages = messages;
        this.missingValueStrategy = missingValueStrategy;
        this.nullValueStrategy = nullValueStrategy;
        this.writer = writer;
    }
    
    public boolean isEscapeXmlEnabled() {
        return escapeXmlEnabled;
    }

    public void setEscapeXmlEnabled(boolean escapeXmlEnabled) {
        this.escapeXmlEnabled = escapeXmlEnabled;
    }

    /**
     * Escapes characters that could be interpreted as XML markup if enabled.
     * 
     * @param input
     *            input string
     * @return escaped string, or the input string if escaping is disabled.
     */
    public String escapeXml(String input) {
        return escapeXmlEnabled ? EscapeXml.escape(input) : input;
    }
    
    /**
     * Resolves variable name to value.
     * 
     * @param key
     *            variable name
     * @return value
     */
    public Object getValue(String key) {
        return model.getValue(key);
    }

    /**
     * Adds a nested scope to search in subsequent lookups.
     * 
     * @param data
     *            data object
     */
    public void pushScope(Object data) {
        model.pushScope(data);
    }
    
    /**
     * Removes innermost nested scope.
     */
    public void popScope() {
        model.popScope();
    }
    
    /**
     * Gets model value that is intended for display in the rendered output.
     * Applies configured strategies for handling missing and null values.
     * 
     * @param modelKey
     *            variable name
     * @return value
     */
    public Object getDisplayValue(String modelKey) {
        Object value = model.getValue(modelKey);
        if (value == Model.NOT_FOUND) {
            value = missingValueStrategy.getValue(modelKey);
        }
        if (value == null) {
            value = nullValueStrategy.getValue(modelKey);
        }
        return value;
    }
    
    /**
     * Gets message pattern from resource bundle.
     * 
     * @param messageKey
     *            message key
     * @return message pattern
     */
    public String getMessage(String messageKey) {
        if (messages == null) {
            throw new TemplateException(
                    "Failed to get message for key '" + messageKey +
                    "' because there is no resource bundle");
        }
        return messages.getString(messageKey);
    }
    
    /**
     * Writes rendered output.
     * 
     * @param str
     *            string to write
     * @return this object
     */
    public RenderContext write(CharSequence str) {
        try {
            writer.append(str);
        } catch (IOException e) {
            throw new RuntimeException("append", e);
        }
        
        return this;
    }
}
